File: logistic_map.py
#!/usr/bin/python2.7
# -*- coding: utf-8 -*-

import numpy as np
import pylab as pl


r_range = np.linspace(0,4,500)

N_pre = 400
N = 10
x = np.zeros(( N, 1 ))
x_init = 0.5


f = lambda x,r : r*x*(1-x)

for r in r_range:
	x[0,0] = x_init
	
	for ite in range(0,N_pre+1):
		x[0,0] = f(x[0,0] , r)
		
		for i in range(0,N-1):
			x[i+1,0] = f(x[i,0] , r) 
	
	pl.scatter(r*np.ones((N,1)), x, 0.2, color='blue')

pl.xlabel('r', fontsize = 20)
pl.ylabel('$x_n$', fontsize =20)
pl.axis((r_range[0],r_range[-1],0,1))
pl.show()

File: logistic_map.m
Npre = 5000; Nplot = 300;
x = zeros(Nplot,1);
for r = 2.5:0.0001:3.7,
  x(1) = 0.5;
  for n = 1:Npre,
    x(1) = r*x(1)*(1 - x(1));
  end,
  for n = 1:Nplot-1,
    x(n+1) = r*x(n)*(1 - x(n));
  end,
  plot(r*ones(Nplot,1), x, '.', 'markersize', 2);
  hold on;
end,
set(gca,'FontSize',25)
title('Bifurcation diagram of the logistic map');
xlabel('r');  ylabel('x_n');
r1 = 2.998;
r2 = 3.449;
r3 = 3.544;
r4 = 3.564;
r5=3.569;

line([r1 r1],[0.2 1],'Color','r','LineWidth',0.2)
line([r2 r2],[0.2 1],'Color','r','LineWidth',0.2)
line([r3 r3],[0.2 1],'Color','r','LineWidth',0.2)
line([r4 r4],[0.2 1],'Color','r','LineWidth',0.2)
line([r5 r5],[0.2 1],'Color','r','LineWidth',0.2)
set(gca, 'xlim');
axis([2.5, 3.7, 0.2 , 1])
hold off; 

r = 0:0.001:4;
a = length(r);

lyap=zeros(1,N);
j=0;
for(r=0:0.001:4)
    xn1=rand(1);
    lyp=0;
    j=j+1;
    for(i=1:10000)
        xn=xn1;
        %logistic map
        xn1=r*xn*(1-xn);
       %wait for transient
       if(i>300)
           % calculate teh sum of logaritm
           lyp=lyp+log(abs(r-2*r*xn1));
       end
    end
    %calculate lyapun
    lyp=lyp/10000;
    lyap(j)=lyp;
end
r = 0:0.001:4;
plot(r,lyap);
